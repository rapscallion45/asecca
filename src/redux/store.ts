import { configureStore, Action } from '@reduxjs/toolkit';
import { ThunkAction } from 'redux-thunk';
import logger from 'redux-logger';
import userPermissionReducer from './slices/userPermissionSlice';
import costsConfigReducer from './slices/costsConfigSlice';
import notificationsReducer from './slices/notificationsSlice';
import themeReducer from './slices/themeSlice';
import { collectionsKanbanSlice } from './slices/kanbanSlice';

/**
 * Configuration of the redux store to manage the global application state
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @see See [more info on Redux Store object](https://redux.js.org/api/store)
 *
 * @constant
 * @type {ToolkitStore<Object>}
 */
const store = configureStore({
  reducer: {
    userPermission: userPermissionReducer,
    costsConfig: costsConfigReducer,
    notifications: notificationsReducer,
    theme: themeReducer,
    collectionsKanban: collectionsKanbanSlice.reducer,
  },
  // @ts-ignore
  middleware:
    process.env.NODE_ENV === 'development'
      ? (getDefaultMiddleware) => getDefaultMiddleware().concat(logger)
      : (getDefaultMiddleware) => getDefaultMiddleware(),
});

/**
 * Shorthand application store dispatch type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppDispatch - application specific store dispatch type
 */
export type AppDispatch = typeof store.dispatch;

/**
 * Shorthand application state type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppState - application specific state object shape
 */
export type AppState = ReturnType<typeof store.getState>;

/**
 * Shorthand application Thunk Action type helper
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef AppThunk - application specific Thunk action definition
 */
export type AppThunk = ThunkAction<void, AppState, unknown, Action<string>>;

export default store;
