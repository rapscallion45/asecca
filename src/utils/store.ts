/* redux store configuration placeholder */
const useStore = () => {};
export default useStore;
