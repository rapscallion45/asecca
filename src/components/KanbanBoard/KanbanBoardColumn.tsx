import React, { FC } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Box, Typography } from '@mui/material';
import CircleIcon from '@mui/icons-material/Circle';
import { dragTask } from '@/redux/slices/kanbanSlice';
import { IKanbanBoardColumn, IKanbanBoardTask } from '@/lib/api/api-types';
import { AppState } from '@/redux/store';
import Task from './KanbanBoardTask';

/**
 * Kanban Board Column Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IKanbanBoardColumnProps
 * @prop {number} colIndex - column index
 */
interface IKanbanBoardColumnProps {
  colIndex: number;
}

/**
 * Kanban Board Column
 *
 * Kanban board column interface for displaying tasks in a certain category
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @component
 * @param {IKanbanBoardColumnProps} props - component props
 * @returns {FC} - data table functional component
 */
const KanbanBoardColumn: FC<IKanbanBoardColumnProps> = (props) => {
  const { colIndex } = props;
  const dispatch = useDispatch();
  const { data: kanbanData } = useSelector((state: AppState) => state.kanban);
  const colColors = ['secondary', 'warning', 'error', 'info', 'primary'];

  /* find this column in board state data from passed column index */
  const column = kanbanData.boards
    .find((board) => board.isActive === true)
    ?.columns.find(
      (col: IKanbanBoardColumn, index: number) => index === colIndex
    );

  /**
   * Callback listens for drop of task into column after drag
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.1
   *
   * @method
   * @param {any} event - object change event
   */
  const handleOnDrop = (event: any) => {
    const { prevColIndex, taskIndex } = JSON.parse(
      event.dataTransfer.getData('text')
    );

    if (colIndex !== prevColIndex) {
      dispatch(dragTask({ colIndex, prevColIndex, taskIndex }));
    }
  };

  /**
   * Callback listens for task drag over column
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.1
   *
   * @method
   * @param {any} event - object change event
   */
  const handleOnDragOver = (event: any) => {
    event.preventDefault();
  };

  /* if no column found in state datam return null */
  return column ? (
    <Box onDrop={handleOnDrop} onDragOver={handleOnDragOver} mr={2}>
      <Box display="flex" flexDirection="row" my={2}>
        {/* @ts-ignore */}
        <CircleIcon fontSize="small" color={colColors[colIndex]} />
        <Typography sx={{ ml: 1 }}>
          {column.name} ({column.tasks.length})
        </Typography>
      </Box>

      {column.tasks.map((task: IKanbanBoardTask, index: number) => (
        <Task key={task.title} taskIndex={index} colIndex={colIndex} />
      ))}
    </Box>
  ) : null;
};

export default KanbanBoardColumn;
