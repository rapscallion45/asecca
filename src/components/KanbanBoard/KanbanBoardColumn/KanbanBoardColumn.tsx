import { FC } from 'react';
import { useSelector } from 'react-redux';
import { Box, Typography } from '@mui/material';
import CircleIcon from '@mui/icons-material/Circle';
import { IKanbanBoardTask } from '@/lib/api/api-types';
import { AppState } from '@/redux/store';
import KanbanBoardTask from '../KanbanBoardTask/KanbanBoardTask';
import { IKanbanBoardColumn } from '../types';

/**
 * Kanban Board Column Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IKanbanBoardColumnProps
 * @prop {number} colIndex - column index
 * @prop {Array<IKanbanBoardColumn>} columns - column data for board on which this column is on
 */
interface IKanbanBoardColumnProps {
  colIndex: number;
  columns: Array<IKanbanBoardColumn>;
}

/**
 * Kanban Board Column
 *
 * Kanban board column interface for displaying tasks in a certain category
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @component
 * @param {IKanbanBoardColumnProps} props - component props
 * @returns {FC} - data table functional component
 */
const KanbanBoardColumn: FC<IKanbanBoardColumnProps> = (props) => {
  const { colIndex, columns } = props;
  const colColors = ['secondary', 'warning', 'error', 'info', 'primary'];
  const { data: tasks } = useSelector((state: AppState) => state.kanban);

  /* find this column in board state data from passed column index */
  const column = columns.find(
    (col: IKanbanBoardColumn, index: number) => index === colIndex
  );

  /* if no column found in state datam return null */
  return column ? (
    <Box mr={2} sx={{ minWidth: 275 }}>
      <Box display="flex" flexDirection="row" my={2}>
        {/* @ts-ignore */}
        <CircleIcon fontSize="small" color={colColors[colIndex]} />
        <Typography sx={{ ml: 1 }}>
          {column.name} (
          {
            tasks.filter(
              (task: IKanbanBoardTask) => task.status === column.name
            ).length
          }
          )
        </Typography>
      </Box>
      {tasks
        .filter((task: IKanbanBoardTask) => task.status === column.name)
        .map((task: IKanbanBoardTask) => (
          <KanbanBoardTask key={task.id} taskId={task.id} columns={columns} />
        ))}
    </Box>
  ) : null;
};

export default KanbanBoardColumn;
