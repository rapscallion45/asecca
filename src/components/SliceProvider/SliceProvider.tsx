import { FC, ReactNode, createContext, useContext } from 'react';
import { useSelector } from 'react-redux';
import { Slice } from '@reduxjs/toolkit';
import { AppState } from '@/redux/store';

/**
 * Default value for the slice context
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof SliceProvider
 *
 * @constant
 * @type {Slice<Object>}
 */
const SliceContext = createContext<Slice>({} as Slice);

/**
 * Slice Provider Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @typedef ISliceProviderProps
 * @prop {Slice} slice - context slice to provide down to child nodes
 * @prop {ReactNode} children - child nodes
 */
interface ISliceProviderProps {
  slice: Slice;
  children: ReactNode;
}

/**
 * Slice Provider
 *
 * Slice provider functional component wrapper for providing global state
 * managment functionality to child components
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 *
 * @component
 * @param {ISliceProviderProps} props - component props
 * @returns {FC} - slice provider functional component
 */
const SliceProvider: FC<ISliceProviderProps> = (props) => {
  const { slice, children } = props;
  return (
    <SliceContext.Provider value={slice}>{children}</SliceContext.Provider>
  );
};

/**
 * Provider specific actions interface for passed slice
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof KanbanReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {CaseReducerActions<any, string>}
 */
export const useSliceActions = () => useContext<Slice>(SliceContext).actions;

/**
 * Provider specific selector interface for passed slice
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.8
 * @memberof KanbanReduxSlice
 *
 * @see See [more info on Redux Slice creation](https://redux-toolkit.js.org/api/createSlice)
 *
 * @constant
 * @type {AppState}
 */
export const useSliceSelector = () => {
  const { name } = useContext<Slice>(SliceContext);
  return useSelector((state: AppState) => state[name as keyof AppState]);
};

export default SliceProvider;
