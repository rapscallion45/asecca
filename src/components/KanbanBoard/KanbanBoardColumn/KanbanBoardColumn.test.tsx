import React from 'react';
import { Provider } from 'react-redux';
import { act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import {
  collectionsKanbanSlice,
  fetchByProjectId as fetchKanbanBoardByProjectId,
} from '@/redux/slices/collectionsKanbanSlice';
import store from '@/redux/store';
import KanbanBoardColumn from './KanbanBoardColumn';
import SliceProvider from '../../SliceProvider/SliceProvider';

/* default test query ID */
const query: string = '66135000001760012';

/* test column index */
const testIdx: number = 0;

/**
 * Kanban Board Column Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Kanban Board Column', () => {
  it('Renders correctly', async () => {
    /** Arrange */
    act(() => {
      store.dispatch(
        fetchKanbanBoardByProjectId({
          projectId: query,
        })
      );
    });

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoardColumn colIndex={testIdx} />
          </SliceProvider>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
