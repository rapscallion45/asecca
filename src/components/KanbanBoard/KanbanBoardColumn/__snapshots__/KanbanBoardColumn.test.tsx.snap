// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kanban Board Column Renders correctly 1`] = `null`;
